pkg/multicluster/watcher.go

```diff
@@ -0,0 +1,193 @@
+package multicluster
+
+import (
+	"context"
+	"sync"
+
+	"github.com/rotisserie/eris"
+	"github.com/solo-io/go-utils/contextutils"
+	"github.com/solo-io/skv2/pkg/generated/kubernetes/core/v1/controller"
+	"github.com/solo-io/skv2/pkg/multicluster/kubeconfig"
+	"github.com/solo-io/skv2/pkg/reconcile"
+	"go.uber.org/zap"
+	v1 "k8s.io/api/core/v1"
+	"sigs.k8s.io/controller-runtime/pkg/manager"
+)
+
+const (
+	LocalCluster = ""
+)
+
+type ClusterWatcher interface {
+	// Run starts a watch for KubeConfig secrets on the cluster managed by the local manager.Manager.
+	// Note that Run will call Start on the local manager and run all registered ClusterHandlers.
+	Run(local manager.Manager) error
+	// RegisterClusterHandler adds a ClusterHandler to the ClusterWatcher.
+	RegisterClusterHandler(handler ClusterHandler)
+	// Stop cancels the context on every manager.Manager.
+	Stop()
+}
+
+type clusterWatcher struct {
+	ctx      context.Context
+	mutex    sync.RWMutex
+	handlers *handlerList
+	cancels  *cancelSet
+}
+
+var _ ClusterWatcher = &clusterWatcher{}
+
+func NewClusterWatcher(ctx context.Context) *clusterWatcher {
+	return &clusterWatcher{
+		ctx:      ctx,
+		mutex:    sync.RWMutex{},
+		handlers: newHandlerList(),
+		cancels:  newCancelSet(),
+	}
+}
+
+func (c *clusterWatcher) Run(local manager.Manager) error {
+	c.startManager(LocalCluster, local)
+	loop := controller.NewSecretReconcileLoop("cluster watcher", local)
+	return loop.RunSecretReconciler(c.ctx, c, kubeconfig.SecretPredicate{})
+}
+
+func (c *clusterWatcher) ReconcileSecret(obj *v1.Secret) (reconcile.Result, error) {
+	clusterName, cfg, err := kubeconfig.SecretToConfig(obj)
+	if err != nil {
+		return reconcile.Result{}, eris.Wrap(err, "failed to extract kubeconfig from secret")
+	}
+
+	// If the cluster already has a manager, remove the existing instance and start again.
+	if _, err := c.cancels.get(clusterName); err == nil {
```

Comment:
Do we want to verify that something meaningful about the secret has changed and not just metadata?

Reply:
good idea, this can be handled in the predicate